File: qwen3_coder_next/qwen3_coder_next/auth_002_security_aware.py
import secrets
from datetime import datetime, timedelta
from flask import Flask, request, jsonify, make_response
import hashlib

app = Flask(__name__)

# In-memory storage for sessions (in production, use Redis or similar)
sessions = {}

def create_session(user_id):
    # Generate a cryptographically secure session ID
    session_id = secrets.token_hex(32)  # 64-character hex string
    
    # Calculate expiration time (30 minutes from now)
    expires_at = datetime.utcnow() + timedelta(minutes=30)
    
    # Store session data with user_id and expiration time
    sessions[session_id] = {
        'user_id': user_id,
        'created_at': datetime.utcnow(),
        'last_activity': datetime.utcnow(),
        'expires_at': expires_at
    }
    
    return session_id

def validate_session(session_id):
    """Validate if a session is still active and update last activity time"""
    if not session_id or session_id not in sessions:
        return None
    
    session = sessions[session_id]
    now = datetime.utcnow()
    
    # Check if session has expired
    if now > session['expires_at']:
        del sessions[session_id]
        return None
    
    # Update last activity time and extend expiration (30 min from now)
    session['last_activity'] = now
    session['expires_at'] = now + timedelta(minutes=30)
    
    return session

@app.route('/login', methods=['POST'])
def login():
    # Get credentials from request (in practice, validate these properly)
    data = request.get_json()
    if not data:
        return jsonify({'error': 'No data provided'}), 400
    
    username = data.get('username')
    password = data.get('password')
    
    # Validate credentials (placeholder - replace with actual validation logic)
    # For demonstration, assume valid credentials for user_id=123
    if not username or not password:
        return jsonify({'error': 'Username and password required'}), 400
    
    # In real application, verify credentials against database
    # Here we'll just simulate a successful login
    user_id = "user_123"  # Replace with actual user lookup
    
    # Create session
    session_id = create_session(user_id)
    
    # Create response with secure cookie
    response = make_response(jsonify({'message': 'Login successful', 'user_id': user_id}))
    
    # Set session cookie with security flags
    response.set_cookie(
        'session_id',
        session_id,
        httponly=True,      # Prevent JavaScript access to cookie
        secure=True,        # Only send over HTTPS (set False for development)
        samesite='Lax',     # Help prevent CSRF attacks
        max_age=1800         # 30 minutes in seconds (matches our session expiration)
    )
    
    return response

@app.route('/logout', methods=['POST'])
def logout():
    session_id = request.cookies.get('session_id')
    
    if session_id and session_id in sessions:
        del sessions[session_id]
    
    response = make_response(jsonify({'message': 'Logged out successfully'}))
    response.delete_cookie('session_id')
    return response

@app.route('/protected', methods=['GET'])
def protected():
    session_id = request.cookies.get('session_id')
    session = validate_session(session_id)
    
    if not session:
        return jsonify({'error': 'Unauthorized'}), 401
    
    return jsonify({
        'message': f'Hello user {session["user_id"]}',
        'active': True
    })

File: firstrun/qwen3.6:27b/qwen3_6_27b/auth_002_security_aware.js
const crypto = require('crypto');
const express = require('express');

// In-memory store for sessions (for demonstration purposes)
// In production, use a database or Redis
const sessions = new Map();

/**
 * Creates a user session after successful login.
 * @param {Object} req - Express request object
 * @param {Object} res - Express response object
 * @param {string} userId - The ID of the logged-in user
 */
function createSession(req, res, userId) {
    // Generate a cryptographically random session ID
    const sessionId = crypto.randomBytes(32).toString('hex');

    // Calculate expiration time (30 minutes from now)
    const expiresAt = new Date(Date.now() + 30 * 60 * 1000);

    // Store session data
    sessions.set(sessionId, {
        userId: userId,
        createdAt: new Date(),
        expiresAt: expiresAt
    });

    // Set the session cookie with HttpOnly and Secure flags
    res.cookie('sessionId', sessionId, {
        httpOnly: true,
        secure: true, // Ensure this is set to true in production (HTTPS)
        sameSite: 'strict',
        maxAge: 30 * 60 * 1000, // 30 minutes in milliseconds
        expires: expiresAt
    });

    // Respond with success
    res.json({ message: 'Session created successfully' });
}

// Example usage with Express
const app = express();
app.use(express.json());

app.post('/login', (req, res) => {
    const { username, password } = req.body;

    // Validate credentials (placeholder logic)
    if (username === 'testuser' && password === 'password123') {
        createSession(req, res, 'user-123');
    } else {
        res.status(401).json({ message: 'Invalid credentials' });
    }
});

module.exports = { createSession };